File: go/periodic_hook.go
package headgate

import "context"

// PeriodicEnqueueAttempt is an immutable snapshot of one durable schedule tick. Accessors
// return owned copies so a hook cannot alter the schedule, tick ID, unique key, or request.
type PeriodicEnqueueAttempt struct {
	schedule ScheduleEntry
	tickMs   int64
	envelope Envelope
}

func newPeriodicEnqueueAttempt(
	schedule ScheduleEntry,
	tickMs int64,
	envelope Envelope,
) PeriodicEnqueueAttempt {
	return PeriodicEnqueueAttempt{
		schedule: cloneScheduleEntry(schedule),
		tickMs:   tickMs,
		envelope: cloneEnqueueBatch([]Envelope{envelope})[0],
	}
}

// Schedule returns an owned copy of the durable schedule.
func (a PeriodicEnqueueAttempt) Schedule() ScheduleEntry { return cloneScheduleEntry(a.schedule) }

// ScheduleID returns the durable schedule identifier.
func (a PeriodicEnqueueAttempt) ScheduleID() string { return a.schedule.ID }

// TickMs returns the scheduled tick in Unix milliseconds.
func (a PeriodicEnqueueAttempt) TickMs() int64 { return a.tickMs }

// Envelope returns an owned copy of the job produced for this tick.
func (a PeriodicEnqueueAttempt) Envelope() Envelope {
	return cloneEnqueueBatch([]Envelope{a.envelope})[0]
}

func cloneScheduleEntry(schedule ScheduleEntry) ScheduleEntry {
	cloned := schedule
	cloned.Payload = cloneBytesPreservingNil(schedule.Payload)
	if schedule.LastEnqueued != nil {
		last := *schedule.LastEnqueued
		cloned.LastEnqueued = &last
	}
	return cloned
}

// PeriodicEnqueueHookPhase identifies whether a hook runs before or after enqueue.
type PeriodicEnqueueHookPhase string

// Periodic enqueue hook phases.
const (
	PeriodicEnqueueHookBegin PeriodicEnqueueHookPhase = "begin"
	PeriodicEnqueueHookEnd   PeriodicEnqueueHookPhase = "end"
)

// PeriodicEnqueueHookEvent surrounds one actual Store enqueue from the scheduler duty.
type PeriodicEnqueueHookEvent struct {
	phase   PeriodicEnqueueHookPhase
	attempt PeriodicEnqueueAttempt
	outcome *InsertOutcome
}

// Phase returns the hook phase.
func (e PeriodicEnqueueHookEvent) Phase() PeriodicEnqueueHookPhase { return e.phase }

// Attempt returns the immutable schedule enqueue attempt.
func (e PeriodicEnqueueHookEvent) Attempt() PeriodicEnqueueAttempt { return e.attempt }

// Outcome returns the store outcome for an end event.
func (e PeriodicEnqueueHookEvent) Outcome() (InsertOutcome, bool) {
	if e.outcome == nil {
		return InsertOutcome{}, false
	}
	return *e.outcome, true
}

// PeriodicEnqueueHook is a synchronous, schedule-aware observer. It cannot mutate or
// replace the durable tick request or Store result.
type PeriodicEnqueueHook interface {
	OnPeriodicEnqueue(context.Context, PeriodicEnqueueHookEvent)
}

// PeriodicEnqueueHookFunc adapts a function to PeriodicEnqueueHook.
type PeriodicEnqueueHookFunc func(context.Context, PeriodicEnqueueHookEvent)

// OnPeriodicEnqueue calls f with the hook event.
func (f PeriodicEnqueueHookFunc) OnPeriodicEnqueue(
	ctx context.Context,
	event PeriodicEnqueueHookEvent,
) {
	f(ctx, event)
}

File: crates/headgate/src/periodic_hook.rs
//! Schedule-aware observations around durable tick enqueue.

use headgate_core::{Envelope, Schedule};

use crate::{InsertOutcome, StoreError};

/// Immutable identity and request data for one durable schedule tick.
#[derive(Clone, Copy, Debug)]
pub struct PeriodicEnqueueAttempt<'a> {
    schedule: &'a Schedule,
    tick_ms: i64,
    envelope: &'a Envelope,
}

impl<'a> PeriodicEnqueueAttempt<'a> {
    pub(crate) fn new(schedule: &'a Schedule, tick_ms: i64, envelope: &'a Envelope) -> Self {
        Self {
            schedule,
            tick_ms,
            envelope,
        }
    }

    pub fn schedule(self) -> &'a Schedule {
        self.schedule
    }

    pub fn schedule_id(self) -> &'a str {
        &self.schedule.id
    }

    pub fn tick_ms(self) -> i64 {
        self.tick_ms
    }

    pub fn envelope(self) -> &'a Envelope {
        self.envelope
    }
}

/// One point event around the exact Store enqueue used by the elected scheduler duty.
#[derive(Clone, Copy, Debug)]
pub enum PeriodicEnqueueHookEvent<'a> {
    Begin {
        attempt: PeriodicEnqueueAttempt<'a>,
    },
    End {
        attempt: PeriodicEnqueueAttempt<'a>,
        outcome: InsertOutcome<'a>,
    },
}

impl<'a> PeriodicEnqueueHookEvent<'a> {
    pub fn attempt(self) -> PeriodicEnqueueAttempt<'a> {
        match self {
            Self::Begin { attempt } | Self::End { attempt, .. } => attempt,
        }
    }

    pub fn outcome(self) -> Option<InsertOutcome<'a>> {
        match self {
            Self::Begin { .. } => None,
            Self::End { outcome, .. } => Some(outcome),
        }
    }
}

/// Synchronous observer for durable periodic enqueue. It cannot mutate the schedule,
/// tick identity, unique key, request, or Store result.
pub trait PeriodicEnqueueHook: Send + Sync + 'static {
    fn on_periodic_enqueue(&self, event: PeriodicEnqueueHookEvent<'_>);
}

/// Function adapter for lightweight periodic observers.
pub struct PeriodicEnqueueHookFn<F>(F);

impl<F> PeriodicEnqueueHookFn<F> {
    pub fn new(function: F) -> Self {
        Self(function)
    }
}

impl<F> PeriodicEnqueueHook for PeriodicEnqueueHookFn<F>
where
    F: for<'a> Fn(PeriodicEnqueueHookEvent<'a>) + Send + Sync + 'static,
{
    fn on_periodic_enqueue(&self, event: PeriodicEnqueueHookEvent<'_>) {
        (self.0)(event);
    }
}

pub(crate) fn outcome_of(result: &Result<(), StoreError>) -> InsertOutcome<'_> {
    InsertOutcome::from_store_result(result)
}
